src/network/serverpackethandler.cpp

```diff
@@ -1501,7 +1501,33 @@ void Server::handleCommand_InventoryFields(NetworkPacket* pkt)
 		return;
 	}
 
-	m_script->on_playerReceiveFields(playersao, formname, fields);
+	if (formname.empty()) { // pass through inventory submits
+		m_script->on_playerReceiveFields(playersao, formname, fields);
+		return;
+	}
+
+	// verify that we displayed the formspec to the user
+	auto peer_state_iterator = m_formspec_state_data.find(pkt->getPeerId());
+	if (peer_state_iterator != m_formspec_state_data.end()) {
+		PeerFormspecState peer_state  = peer_state_iterator->second;
+		if (formname == peer_state.getClientSideName()) {
+			m_script->on_playerReceiveFields(playersao, peer_state.formspec_name, fields);
+			if (fields["quit"] == "true")
+				m_formspec_state_data.erase(peer_state_iterator);
+			return;
+		} else {
```

Comment:
Superfluous `else`. Case above always returns.

Reply:
seemed cleaner to write it that way but I can change it